standards/HEALTH_CHECK_SPECIFICATION.md

```diff
@@ -0,0 +1,101 @@
+
+Some examples of things that can be checked are:
+
+* the status of the connections to the data stores and other third party dependencies
+* the status of other apps the app depends on
+* the status of the host on which the app is running
+* the status of the application logic (e.g. kafka message processing failures, request processing failures, concurrent request thresholds, etc.)
+
+The following requirements must be met by health check implementations:
+
+* The app must perform checks in order to determine its current status.
+* The checks must all run all on a fixed interval that can be set per app
+* The checks must have a failure threshold per check of the number of sequential failed checks before which the check is marked as a failure
+* The checks must have a healthy threshold per check of the number of sequential healthy checks before which the check is marked as healthy
+* Changes in check statuses (including the overall app status) must be logged out
+
+### Check statuses
+
+The status of the check can be one of three values:
+
+Status     | Description
+-----------|--------------
+`OK`       | Everything is ok
+`WARNING`  | Things are degraded, but at least partially functioning
+`CRITICAL` | Failure, your app will now be unable to function
+
+Health check endpoint
+---------------------
+
+The health check data is served via the health check endpoint.  This health check data comes in two forms.  The status code provides a quick way to get the overall status of the app, while the json body provides further information that might be useful in triaging any issues.
+
+The following requirements must be met by health check implementations:
+
+* All apps must have a health check endpoint
+* The health check endpoint must be available as a `GET /healthcheck` endpoint
+* The health check must return the [appropriate status code](#health-check-status-codes)
+* The health check must return [status information as json](#health-check-body)
+
+### Health check status codes
+
+There are three standard status codes that a health check can return:
+
+Status code | Description
+------------|------------------------
+`200`       | I'm fine
+`429`       | I'm warming up or I'm degraded
+`500`       | Kill me
+
+The health check should return 429 until the first checks have been run.  After that the codes returned should be based on the check status.  It is the responsibility of whatever is monitoring the health check to provide a grace period on startup.
+
+If any check returns a `CRITICAL` status (see [check statuses below](#check-statuses) then the app should transition to a `429`.  After a timeout that can be set per app the `429` will be downgraded `500`.  This delay in marking the app as `500` is to allow the issue a chance to resolve on its own before moving to a `500` which will result in the app being killed.
```

Comment:
I am not sure I follow this logic. My understanding is when any health check fails *internally* its status will transition from `OK` to `DEGRADED` this starts a _doomsday counter_ which is a timeout for the app to pull its self together. While this is happening the endpoint will return HTTP `429` unless the app recovers it self. If the countdown expires and the app is still borked then its internal status is `CRITICAL` and it returns a 500.

Question: Should an app ever recover from a `CRITICAL` status? Or do we rely on the platform at that point to kill it and start it up again.

Reply:
After a conversation with Jon I now realise that `check` here means the health check function within the app not the endpoint.

Reply:
@daiLlew Potentially adding clarity of the terminology (i.e. check) would be helpful then?

Reply:
I added a little bit of a description to the list of components at the top of the spec.  Hopefully this is helpful.